pkg/runner/runner.go

```diff
@@ -80,39 +92,59 @@ func NewFromWatches(path string) (map[schema.GroupVersionKind]Runner, error) {
 }
 
 // NewForPlaybook returns a new Runner based on the path to an ansible playbook.
-func NewForPlaybook(path string, gvk schema.GroupVersionKind) Runner {
-	return &runner{
+func NewForPlaybook(path string, gvk schema.GroupVersionKind, finalizer *Finalizer) (Runner, error) {
+	if !filepath.IsAbs(path) {
+		return nil, fmt.Errorf("playbook path must be absolute for %v", gvk)
+	}
+	r := &runner{
 		Path: path,
 		GVK:  gvk,
 		cmdFunc: func(ident, inputDirPath string) *exec.Cmd {
-			dc := exec.Command("ansible-runner", "-vv", "-p", path, "-i", ident, "run", inputDirPath)
-			return dc
+			return exec.Command("ansible-runner", "-vv", "-p", path, "-i", ident, "run", inputDirPath)
 		},
 	}
+	err := r.addFinalizer(finalizer)
+	if err != nil {
+		return nil, err
+	}
+	return r, nil
 }
 
 // NewForRole returns a new Runner based on the path to an ansible role.
-func NewForRole(path string, gvk schema.GroupVersionKind) Runner {
+func NewForRole(path string, gvk schema.GroupVersionKind, finalizer *Finalizer) (Runner, error) {
+	if !filepath.IsAbs(path) {
+		return nil, fmt.Errorf("role path must be absolute for %v", gvk)
+	}
 	path = strings.TrimRight(path, "/")
-	return &runner{
+	r := &runner{
 		Path: path,
 		GVK:  gvk,
 		cmdFunc: func(ident, inputDirPath string) *exec.Cmd {
 			rolePath, roleName := filepath.Split(path)
-			dc := exec.Command("ansible-runner", "-vv", "--role", roleName, "--roles-path", rolePath, "--hosts", "localhost", "-i", ident, "run", inputDirPath)
-			return dc
+			return exec.Command("ansible-runner", "-vv", "--role", roleName, "--roles-path", rolePath, "--hosts", "localhost", "-i", ident, "run", inputDirPath)
 		},
 	}
+	err := r.addFinalizer(finalizer)
+	if err != nil {
+		return nil, err
+	}
+	return r, nil
 }
 
 // runner - implements the Runner interface for a GVK that's being watched.
 type runner struct {
-	Path    string                                     // path on disk to a playbook or role depending on what cmdFunc expects
-	GVK     schema.GroupVersionKind                    // GVK being watched that corresponds to the Path
-	cmdFunc func(ident, inputDirPath string) *exec.Cmd // returns a Cmd that runs ansible-runner
+	Path             string                  // path on disk to a playbook or role depending on what cmdFunc expects
+	GVK              schema.GroupVersionKind // GVK being watched that corresponds to the Path
+	Finalizer        *Finalizer
+	cmdFunc          func(ident, inputDirPath string) *exec.Cmd // returns a Cmd that runs ansible-runner
+	finalizerCmdFunc func(ident, inputDirPath string) *exec.Cmd
 }
 
 func (r *runner) Run(u *unstructured.Unstructured, kubeconfig string) (chan eventapi.JobEvent, error) {
+	if u.GetDeletionTimestamp() != nil && !r.isFinalizerRun(u) {
+		logger.Warn("Resource has been deleted, but no finalizer was matched, skipping reconciliation")
+		return nil, nil
```

Comment:
This seems weird to me

Reply:
I agree this should probably return an error